### ISSUE TITLE
Sometimes, exceptions cannot be logged because "DefaultTransactionContext is not marked as serializable."

### ISSUE THREAD

**Issue by Rzpeg** [rebus-org/Rebus#493]
Hello :smile: 
In very rare cases, when an unhandled exception occures, Rebus can't log the error because of the following: 

`Unhandled exception {0} while handling message with ID {1}: {2}" Argument0="1" Argument1="847a71b3-e8ca-4c9a-86b2-7ebca84bb086" Argument2="System.Runtime.Serialization.SerializationException: Type 'Rebus.Transport.DefaultTransactionContext' in assembly 'Rebus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.`

Any ideas ?

**Comment by mookid8000** [comment]
Any chance you can post the full stack trace?

**Comment by Rzpeg** [comment]
Sure:
```
Unhandled exception {0} while handling message with ID {1}: {2}" Argument0="1" Argument1="847a71b3-e8ca-4c9a-86b2-7ebca84bb086" Argument2="System.Runtime.Serialization.SerializationException: Type 'Rebus.Transport.DefaultTransactionContext' in assembly 'Rebus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.
   at System.AppDomain.get_Evidence()
   at System.AppDomain.get_Evidence()
   at System.Configuration.ClientConfigPaths.GetEvidenceInfo(AppDomain appDomain, String exePath, String& typeName)
   at System.Configuration.ClientConfigPaths.GetTypeAndHashSuffix(AppDomain appDomain, String exePath)
   at System.Configuration.ClientConfigPaths..ctor(String exePath, Boolean includeUserConfig)
   at System.Configuration.ClientConfigPaths.GetPaths(String exePath, Boolean includeUserConfig)
   at System.Configuration.ClientConfigurationHost.RequireCompleteInit(IInternalConfigRecord record)
   at System.Configuration.BaseConfigurationRecord.GetSectionRecursive(String configKey, Boolean getLkg, Boolean checkPermission, Boolean getRuntimeObject, Boolean requestIsHere, Object& result, Object& resultRuntimeObject)
   at System.Configuration.BaseConfigurationRecord.GetSection(String configKey)
   at System.Configuration.ConfigurationManager.GetSection(String sectionName)
   at System.Xml.XmlConfiguration.XmlReaderSection.CreateDefaultResolver()
   at System.Xml.Schema.XmlSchema.Read(XmlReader reader, ValidationEventHandler validationEventHandler)
   at System.Data.Entity.Core.SchemaObjectModel.Schema.SomSchemaSetHelper.AddXmlSchemaToSet(XmlSchemaSet schemaSet, XmlSchemaResource schemaResource, HashSet`1 schemasAlreadyAdded)
   at System.Data.Entity.Core.SchemaObjectModel.Schema.SomSchemaSetHelper.AddXmlSchemaToSet(XmlSchemaSet schemaSet, XmlSchemaResource schemaResource, HashSet`1 schemasAlreadyAdded)
   at System.Data.Entity.Core.SchemaObjectModel.Schema.SomSchemaSetHelper.ComputeSchemaSet(SchemaDataModelOption dataModel)
   at System.Data.Entity.Core.Common.Utils.Memoizer`2.Result.GetValue()
   at System.Data.Entity.Core.Common.Utils.Memoizer`2.Evaluate(TArg arg)
   at System.Data.Entity.Core.SchemaObjectModel.Schema.SomSchemaSetHelper.GetSchemaSet(SchemaDataModelOption dataModel)
   at System.Data.Entity.Core.SchemaObjectModel.Schema.Parse(XmlReader sourceReader, String sourceLocation)
   at System.Data.Entity.Core.SchemaObjectModel.SchemaManager.ParseAndValidate(IEnumerable`1 xmlReaders, IEnumerable`1 sourceFilePaths, SchemaDataModelOption dataModel, AttributeValueNotification providerNotification, AttributeValueNotification providerManifestTokenNotification, ProviderManifestNeeded providerManifestNeeded, IList`1& schemaCollection)
   at System.Data.Entity.Core.Metadata.Edm.EdmItemCollection.LoadItems(IEnumerable`1 xmlReaders, IEnumerable`1 sourceFilePaths, SchemaDataModelOption dataModelOption, DbProviderManifest providerManifest, ItemCollection itemCollection, Boolean throwOnError)
   at System.Data.Entity.Core.Metadata.Edm.MetadataCache.LoadEdmItemCollection(MetadataArtifactLoader loader)
   at System.Data.Entity.Core.Metadata.Edm.MetadataCache.<>c__DisplayClass5.<GetMetadataWorkspace>b__0(String k)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at System.Data.Entity.Core.Metadata.Edm.MetadataCache.GetMetadataWorkspace(String cacheKey, MetadataArtifactLoader artifactLoader)
   at System.Data.Entity.Core.EntityClient.EntityConnection.GetMetadataWorkspace()
   at System.Data.Entity.Core.Objects.ObjectContext..ctor(EntityConnection connection, Boolean isConnectionConstructor, ObjectQueryExecutionPlanFactory objectQueryExecutionPlanFactory, Translator translator, ColumnMapFactory columnMapFactory)
   at System.Data.Entity.Internal.InternalConnection.CreateObjectContextFromConnectionModel()
   at System.Data.Entity.Internal.LazyInternalContext.InitializeContext()
   at System.Data.Entity.Internal.InternalContext.GetEntitySetAndBaseTypeForType(Type entityType)
   at System.Data.Entity.Internal.Linq.InternalSet`1.Initialize()
   at System.Data.Entity.Internal.Linq.InternalSet`1.get_InternalContext()
   at System.Data.Entity.Infrastructure.DbQuery`1.System.Linq.IQueryable.get_Provider()
   at System.Data.Entity.QueryableExtensions.FirstOrDefaultAsync[TSource](IQueryable`1 source, Expression`1 predicate, CancellationToken cancellationToken)
   at System.Data.Entity.QueryableExtensions.FirstOrDefaultAsync[TSource](IQueryable`1 source, Expression`1 predicate)
   at TestApp.Handlers.AssignmentPlanner.AssignmentClosedEventHandler.<Handle>d__3.MoveNext() in C:\dev\git\Proto\TestApp\src\TestApp.Handlers.AssignmentPlanner\AssignmentClosedEventHandler.cs:line 45--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Pipeline.Receive.HandlerInvoker`1.<Invoke>d__13.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Pipeline.Receive.DispatchIncomingMessageStep.<Process>d__1.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Sagas.LoadSagaDataStep.<Process>d__7.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Pipeline.Receive.ActivateHandlersStep.<Process>d__3.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Retry.Simple.FailedMessageWrapperStep.<Process>d__2.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Pipeline.Receive.DeserializeIncomingMessageStep.<Process>d__2.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Retry.Simple.SimpleRetryStrategyStep.<DispatchWithTrackerIdentifier>d__8.MoveNext()
```

**Comment by mookid8000** [comment]
As you can see from the stack trace, the exception gets thrown when `AssignmentClosedEventHandler` calls `FirstOrDefaultAsync` on an EF query on line 45, which - apparently! - deep down and very indirectly causes Entity Framework to end up wanting to serialize Rebus' `DefaultTransactionContext` (*)

Rebus stores its ambient transaction context via `CallContext.LogicalSetData`/`CallContext.LogicalGetData`, and for some reason that data seems to be serialized when EF executes a query.

From the stack trace I can gather the following: EF seems to ask for some metadata regarding the model type (`GetMetadataWorkspace`) which seems to go to a concurrent dictionary's `GetOrAdd` method, which in turn seems to call `LoadEdmItemCollection` on a `MetadataCache` which seems to call further down and load some XML configuration from somewhere. In doing this, it ends up trying to call `get` for the app domain's `Evidence`, which is the operation that seems to fail.

The `GetOrAdd` on the concurrent dictionary makes me believe that this is a one-time initialization thing that MAYBE could be initialized instead when your application starts up, to avoid initializing it inside a Rebus message handler.

However, I have no idea how to do that. Maybe you know what this is about?


---

(*) Which - as the exception says - cannot be done, as it has not been marked with the `[Serializable]` attribute - and putting the `[Serializable]` attribute on it is not an option, because it contains a dictionary of things that are needed for Rebus' current message transaction, e.g. database connections etc.

**Comment by mookid8000** [comment]
oh btw I closed the issue because it is not really Rebus' issue.... Entity Framework has some very specific requirements for the thread that uses it, apparently..... if those requirements were stated somewhere and there was something I could do that I would know for sure would solve the problem, I would definitely do it....

**Comment by Rzpeg** [comment]
And again, now with Azure Service Bus. So strange
```

EventName="FormattedMessageEvent" FormattedMessage="Rebus.Config.RebusConfigurer+<>c: Moving message with ID {0} to error queue '{1}' - reason: {2}" Argument0="6cef56ad-5949-4695-abfc-5039599682e4" Argument1="Marlo.M3.WorkerService.error" Argument2="3 unhandled exceptions: 9/7/2016 4:02:01 PM +00:00: System.Configuration.ConfigurationErrorsException: Configuration system failed to initialize ---> System.Runtime.Serialization.SerializationException: Type 'Rebus.Transport.DefaultTransactionContext' in assembly 'Rebus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.
 System.AppDomain.get_Evidence()
 System.AppDomain.get_Evidence()
 System.Configuration.ClientConfigPaths.GetEvidenceInfo(AppDomain appDomain, String exePath, String& typeName)
 System.Configuration.ClientConfigPaths.GetTypeAndHashSuffix(AppDomain appDomain, String exePath)
 System.Configuration.ClientConfigPaths..ctor(String exePath, Boolean includeUserConfig)
 System.Configuration.ClientConfigPaths.GetPaths(String exePath, Boolean includeUserConfig)
 System.Configuration.ClientConfigurationHost.get_HasRoamingConfig()
 System.Configuration.ClientConfigurationHost.IsConfigRecordRequired(String configPath)
 System.Configuration.BaseConfigurationRecord.hlNeedsChildFor(String configName)
 System.Configuration.Internal.InternalConfigRoot.GetConfigRecord(String configPath)
 System.Configuration.ClientConfigurationSystem.OnConfigRemoved(Object sender, InternalConfigEventArgs e)   --- End of inner exception stack trace ---
 System.Configuration.ConfigurationManager.PrepareConfigSystem()
 System.Configuration.ConfigurationManager.get_AppSettings()
 Microsoft.ServiceBus.Messaging.SoapProtocolDefaults.IsAvailableClientCertificateThumbprint(String& thumbprint)9/7/2016 4:02:01 PM +00:00: System.Configuration.ConfigurationErrorsException: Configuration system failed to initialize ---> System.Runtime.Serialization.SerializationException: Type 'Rebus.Transport.DefaultTransactionContext' in assembly 'Rebus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.
 System.AppDomain.get_Evidence()
 System.AppDomain.get_Evidence()
 System.Configuration.ClientConfigPaths.GetEvidenceInfo(AppDomain appDomain, String exePath, String& typeName)
 System.Configuration.ClientConfigPaths.GetTypeAndHashSuffix(AppDomain appDomain, String exePath)
 System.Configuration.ClientConfigPaths..ctor(String exePath, Boolean includeUserConfig)
 System.Configuration.ClientConfigPaths.GetPaths(String exePath, Boolean includeUserConfig)
 System.Configuration.ClientConfigurationHost.get_HasRoamingConfig()
 System.Configuration.ClientConfigurationHost.IsConfigRecordRequired(String configPath)
 System.Configuration.BaseConfigurationRecord.hlNeedsChildFor(String configName)
 System.Configuration.Internal.InternalConfigRoot.GetConfigRecord(String configPath)
 System.Configuration.ClientConfigurationSystem.OnConfigRemoved(Object sender, InternalConfigEventArgs e)   --- End of inner exception stack trace ---
 System.Configuration.ConfigurationManager.PrepareConfigSystem()
 System.Configuration.ConfigurationManager.get_AppSettings()
 Microsoft.ServiceBus.Messaging.SoapProtocolDefaults.IsAvailableClientCertificateThumbprint(String& thumbprint)9/7/2016 4:02:01 PM +00:00: System.Configuration.ConfigurationErrorsException: Configuration system failed to initialize ---> System.Runtime.Serialization.SerializationException: Type 'Rebus.Transport.DefaultTransactionContext' in assembly 'Rebus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.
 System.AppDomain.get_Evidence()
 System.AppDomain.get_Evidence()
 System.Configuration.ClientConfigPaths.GetEvidenceInfo(AppDomain appDomain, String exePath, String& typeName)
 System.Configuration.ClientConfigPaths.GetTypeAndHashSuffix(AppDomain appDomain, String exePath)
 System.Configuration.ClientConfigPaths..ctor(String exePath, Boolean includeUserConfig)
 System.Configuration.ClientConfigPaths.GetPaths(String exePath, Boolean includeUserConfig)
 System.Configuration.ClientConfigurationHost.get_HasRoamingConfig()
 System.Configuration.ClientConfigurationHost.IsConfigRecordRequired(String configPath)
 System.Configuration.BaseConfigurationRecord.hlNeedsChildFor(String configName)
 System.Configuration.Internal.InternalConfigRoot.GetConfigRecord(String configPath)
 System.Configuration.ClientConfigurationSystem.OnConfigRemoved(Object sender, InternalConfigEventArgs e)   --- End of inner exception stack trace ---
 System.Configuration.ConfigurationManager.PrepareConfigSystem()
 System.Configuration.ConfigurationManager.get_AppSettings()
 Microsoft.ServiceBus.Messaging.SoapProtocolDefaults.IsAvailableClientCertificateThumbprint(String& thumbprint)" TraceSource="WaWorkerHost.exe"
```

**Comment by mookid8000** [comment]
@Rzpeg I just had a thought – I seem to recall at some point to have experienced the `Type 'Rebus.Transport.DefaultTransactionContext' in assembly 'Rebus, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.` error in relation to something else... and that time the problem was that something got used the first time on a Rebus worker thread, and thus its first-time initialization would run there.... and for some reason, initializing that thing required all kinds of crazy cross-domain stuff to happen, thus failing with the same error message as this one.

It is not clear to me what is actually failing in the stack trace you have posted – I mean, i shows that the `IsAvailableClientCertificateThumbprint` method started it, but where is that called?

**Comment by Rzpeg** [comment]
the error is "System.Configuration.ConfigurationErrorsException: Configuration system failed to initialize", which is strange as it only happens in worker role production. In azure emulator or as console app it runs OK.

**Comment by mookid8000** [comment]
But the question is: Where does it happen?

As in: Which exact line of code (that either you or I control) experienced that exception.

Maybe I am just tired or something – but I cannot see that from the information you have posted so far.

**Comment by Rzpeg** [comment]
Ok, I think I got it.
It happens in EF 6 first-time initialization engine when there are more than 2 concurrent Rebus worker threads call for EF for the first time in the same time, EF static config initializer fails with a cross-domain call, and never manages to recover itself. What resolved the issue is a dummy query to EF context on app startup, before starting Rebus. Still happens only in Azure worker roles release ambiente though...

**Comment by jr01** [comment]
I just ran into the same issue with Rebus. I can reproduce it locally using a workerrole in Azure emulator.

Here are some details to reproduce: 
```
public class MyNonSerializableClass
    {
        public ConcurrentDictionary<string, object> Items { get; } = new ConcurrentDictionary<string, object>();
    }
```

workerrole.cs
```
public override bool OnStart()
        {
            // Simulate the Rebus worker startup.
            CallContext.LogicalSetData("stuff", new MyNonSerializableClass());

            // Query to trigger entity framework initialization
            MyEntityFrameworkQuery();

            return base.OnStart();
        }
```

The issue is well described here:
https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/mitigation-deserialization-of-objects-across-app-domains

The suggested mitigation, `ConfigurationManager.GetSection("system.xml/xmlReader") ` as first line in the OnStart resolves the issue.  The workaround @Rzpeg described of course also works.

I am not sure if that `GetSection("system.xml/xmlReader")` should be done in Rebus.

**Comment by Rzpeg** [comment]
ping @mookid8000 @jr01
Is it still an issue in Rebus 4 ? Or ConfigurationManager went away with .net standard update?

**Comment by mookid8000** [comment]
> Is it still an issue in Rebus 4 ? Or ConfigurationManager went away with .net standard update?

It is not really an issue with Rebus specifically, it is more an issue with the way `ConfigurationManager` works, in that it works in wonky ways when it is being initialized (as in: requires that nothing un-serializable has been added to the logical call context).

But since this issue seems to plague so many people, I have added to the .NET 4.5 build of Rebus 4.1.0-b05 the aforementioned `GetSection` mitigation.

I don't think this is an issue in the .NET Standard build of Rebus, because there's no configuration manager and no silly juggling of app domains going on 😁 

4.1.0-b05 will be out in a few minutes as soon as the build server has finished munching through all the tests.
